code/drasil-gool/lib/Drasil/GOOL/LanguageRenderer/JuliaRenderer.hs

```diff
@@ -0,0 +1,1064 @@
+  blockCmt, docCmt, commentedMod, listSetFunc, commentedItem, break, continue)
+import Drasil.GOOL.LanguageRenderer.Constructors (mkVal, mkStateVal, VSOp,
+  unOpPrec, powerPrec, unExpr, unExpr', binExpr, multPrec, typeUnExpr,
+  typeBinExpr, mkStmt, mkStmtNoEnd)
+import Drasil.GOOL.LanguageRenderer.LanguagePolymorphic (OptionalSpace(..))
+import qualified Drasil.GOOL.LanguageRenderer.LanguagePolymorphic as G (
+  block, multiBlock, listInnerType, litChar, litDouble, litInt, litString,
+  valueOf, negateOp, equalOp, notEqualOp, greaterOp, greaterEqualOp, lessOp,
+  lessEqualOp, plusOp, minusOp, multOp, divideOp, moduloOp, var, call,
+  funcAppMixedArgs, lambda, listAccess, listSet, arrayElem, modFromData,
+  fileDoc, fileFromData, tryCatch, csc, multiBody, sec, cot, stmt, loopStmt,
+  emptyStmt, assign, increment, subAssign, print, comment, valStmt, function,
+  returnStmt, construct, param, docFunc, throw, arg, argsList, ifCond, smartAdd)
+import qualified Drasil.GOOL.LanguageRenderer.CommonPseudoOO as CP (bool,
+  boolRender, extVar, funcType, buildModule, docMod', funcDecDef, litArray,
+  listDec, listDecDef, listAccessFunc, listSetFunc, bindingError, notNull,
+  extFuncAppMixedArgs, functionDoc, listSize, listAdd, listAppend, intToIndex',
+  indexToInt', inOutFunc, docInOutFunc', forLoopError, varDecDef, openFileR',
+  openFileW', openFileA', multiReturn, multiAssign, inOutCall, mainBody,
+  argExists, forEach')
+import qualified Drasil.GOOL.LanguageRenderer.CLike as C (litTrue, litFalse,
+  notOp, andOp, orOp, inlineIf, while)
+import qualified Drasil.GOOL.LanguageRenderer.Macros as M (increment1,
+  decrement1, ifExists, stringListVals, stringListLists)
+import Drasil.GOOL.AST (Terminator(..), FileType(..), FileData(..), fileD,
+  FuncData(..), ModData(..), md, updateMod, MethodData(..), mthd, OpData(..),
+  ParamData(..), ProgData(..), TypeData(..), td, ValData(..), vd, VarData(..),
+  vard, CommonThunk, progD, fd, pd, updateMthd, commonThunkDim, commonThunkElim,
+  vectorize, vectorize2, commonVecIndex, sumComponents, pureValue)
+import Drasil.GOOL.Helpers (vibcat, toCode, toState, onCodeValue, onStateValue,
+  on2CodeValues, on2StateValues, onCodeList, onStateList, emptyIfEmpty)
+import Drasil.GOOL.State (VS, lensGStoFS, revFiles, setFileType, lensMStoVS,
+  getModuleImports, addModuleImportVS, getUsing, getLangImports, getLibImports,
+  addLibImportVS, useVarName, getMainDoc, genLoopIndex, genVarNameIf)
+
+import Control.Applicative (liftA2)
+import Prelude hiding (break,print,sin,cos,tan,floor,(<>))
+import Data.Maybe (fromMaybe, isNothing)
+import Data.Functor ((<&>))
+import Control.Lens.Zoom (zoom)
+import Control.Monad.State (modify)
+import Data.List (intercalate, sort)
+import Text.PrettyPrint.HughesPJ (Doc, text, (<>), (<+>), empty, brackets, vcat,
+  quotes, doubleQuotes, parens, equals, colon)
+import qualified Text.PrettyPrint.HughesPJ as D (float)
+
+jlExt :: String
+jlExt = "jl"
+
+newtype JuliaCode a = JLC {unJLC :: a}
```

Comment:
can probably use `deriving` for `Functor`, `Applicative` and `Monad`.

Reply:
I tried removing the `instance ... JuliaCode` lines for those three and changing line `98` to be

```haskell
newtype JuliaCode a = JLC {unJLC :: a} deriving (Functor, Applicative, Monad)
```

This gave me errors for all 3:
```
    • Can't make a derived instance of ‘Functor JuliaCode’:
        You need DeriveFunctor to derive an instance for this class
    • In the newtype declaration for ‘JuliaCode’
   |
98 | newtype JuliaCode a = JLC {unJLC :: a} deriving (Functor, Applicative, Monad)
   |                                                  ^^^^^^^
```
```
    • Can't make a derived instance of ‘Applicative JuliaCode’:
        ‘Applicative’ is not a stock derivable class (Eq, Show, etc.)
        Try GeneralizedNewtypeDeriving for GHC's newtype-deriving extension
    • In the newtype declaration for ‘JuliaCode’
   |
98 | newtype JuliaCode a = JLC {unJLC :: a} deriving (Functor, Applicative, Monad)
   |                                                           ^^^^^^^^^^^
```
```
    • Can't make a derived instance of ‘Monad JuliaCode’:
        ‘Monad’ is not a stock derivable class (Eq, Show, etc.)
        Try GeneralizedNewtypeDeriving for GHC's newtype-deriving extension
    • In the newtype declaration for ‘JuliaCode’
   |
98 | newtype JuliaCode a = JLC {unJLC :: a} deriving (Functor, Applicative, Monad)
   |                                                                        ^^^^^
```
I added these to the top of the file, as recommended:
```haskell
{-# LANGUAGE DeriveFunctor #-}
{-# LANGUAGE GeneralizedNewtypeDeriving #-}
```
This fixed the `deriving Functor` error, but then for `Applicative` (and `Monad` was similar) I got:
```
    • Can't make a derived instance of ‘Applicative JuliaCode’
        (even with cunning GeneralizedNewtypeDeriving):
        cannot eta-reduce the representation type enough
    • In the newtype declaration for ‘JuliaCode’
    |
100 | newtype JuliaCode a = JLC {unJLC :: a} deriving (Functor, Applicative, Monad)
    |                                                           ^^^^^^^^^^^
```

So it looks like we can use `deriving Functor`, but we need an instance of `Applicative` and `Monad`.

I tested it with `deriving Functor`, and that seemed to work.  Should we stick with that?

Reply:
Odd - I am quite sure this should work. But it might need more 'hints'. Have you looked at the documentation for GeneralizedNewtypeDeriving ?

Reply:
After reading the documentation (and a surprisingly well-explained answer from ChatGPT when I was feeling out of my depth), it appears that the `GeneralizedNewtypeDeriving` language extension only works with `newtype`s that take a basic list of type variables.  Since `JuliaCode` has type variables nested within a record type, it doesn't work.

[Here](https://ghc.gitlab.haskell.org/ghc/doc/users_guide/exts/newtype_deriving.html#a-more-precise-specification) is the specification, if you're interested.

Reply:
Ah, thanks for the investigation.